Fix dipole and network (spiking) class write methods

rythorpe opened the pull request:
Two changes have been made: 1) `dipole.write()` has been fixed to save files according to the user-specified name, and 2) `network.write_spikes()` has been added so that spiking data for selected trials can be saved to a single .txt file. See the attached example (a modified version of the plot_simulate_evoked.py example) and its corresponding output .txt and .png files that demonstrates reconstruction of the simulated spike event plot using the spike .txt files.
[demo.zip](https://github.com/jonescompneurolab/hnn-core/files/4391152/demo.zip)

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +518,32 @@ def plot_spikes(self, ax=None, show=True):
+        for spike_type,gid_range in self.gid_dict.items():
+            gidtypes[np.in1d(spikegids,gid_range)] = spike_type
+
+        with open(fname,'w') as f:
```
you should use standard functions whenever possible. In this case, I think you could use [np.savetxt](https://docs.scipy.org/doc/numpy/reference/generated/numpy.savetxt.html)

rythorpe replied:
Because I combined data types (i.e., list of floats and list of strings) which don't easily combine into a single array, the most succinct way to handle the formatting was to use low-level I/O functions. I could potentially convert each list to a formatted string, then convert it to a numpy array, and then save it if you think that'd be better?

Note that HNN originally stored spike.txt files along side param.txt files that stored the gid ranges. I tried to retain the HNN format but also combine all the relevant spiking info into a single file.

blakecaldwell replied:
Since we are proceeding with using hnn-core for HNN. I'd prefer to keep spikes for different trials in separate files, as with HNN. If it's not, I will have to write a function for HNN that reads the single file and then splits into different files. :(

rythorpe replied:
It is currently setup to take an argument (i.e., `trial_idx`) that specifies which trials are selected to write to the file, so the write function can be looped over to save all trials separately. The reason for this was to mimic the low-level functionality of the `dipole.write()` function which also only writes one file at a time. Would it be more helpful if we changed both the `dipole.write()` and `network.write_spikes()` functions to output all files with the same naming convention as HNN?

blakecaldwell replied:
Oh, yes. This is where a separate Spike (Spikedata) class would be helpful for parity with Dipole: Spikedata.write() and Spikedata.write_all_trials(). That way you can enforce some standard naming (like HNN for better or worse) and don't have the ambiguous case of naming a file with some, but not all trials.

There's more than one way to do it. Just my thoughts. Thanks for your work on this.

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +518,32 @@ def plot_spikes(self, ax=None, show=True):
+        Parameters
+        ----------
+        fname : str
+            Full path to the output file (.txt)
```
What motivated the use of a `.txt` file instead of say an `.npz` or `.mat` file?

rythorpe replied:
I just wanted to be consistent with the `dipole.write()` function and previous HNN I/O convention.

blakecaldwell replied:
Since testing new hnn-core output against old hnn output will be important for the foreseeable future, I plead that we leave files the same (.txt). We will have test functions to compare array values after loading from the file, but also being able to do a `diff` from the command line is useful.

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +518,32 @@ def plot_spikes(self, ax=None, show=True):
+            Indices of selected trials. If 'all',
+            all trials are selected.
+        """
+        if trial_idx=='all':
```
use `None`. The problem with `'all'` is that string comparison may fail if it's not the right case

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +518,32 @@ def plot_spikes(self, ax=None, show=True):
         return ax.get_figure()
+
+    def write_spikes(self, fname, trial_idx='all'):
+        """Write spike times to a file.
```
how do you plan to read them back in?

rythorpe replied:
Check out the plot_simulate_evoked.py script in the attached demo.zip folder. Are you thinking we should have a built-in read function as well?

blakecaldwell replied:
HNN has the functionality to take a given parameter file, find the spike output in an expected directory and load it. If this functionality is part of hnn-core, HNN can use it. Otherwise, the function will still be part of HNN.

rythorpe replied:
So are you in favor of adding a read function in hnn-core or would you rather that be handled by an HNN wrapper?

jasmainak replied:
I think you should add a reader if you're going to have a way to write it :) It's also easier to test that way. You can do a round trip and check if you have what you started with.

rythorpe replied:
Since spikes currently exist only as a `network` class attribute, they either need to be read in through a `network` class method at/after network initialization or handled as their own object outside of the `network` class (as with params). In keeping with our desire to make HNN data structures more transparent and tractable (e.g., using .json files for params, json schema for param validation), what do you think of turning spikes into their own object with the following example structure:
```
{
    <gid_type>: {
        <gid>: [
            <spiketime[0]>,
            <spiketime[1]>,
            ...
            ]
        }
}
```
We could still maintain the .txt file format convention of HNN output for now, but this would make the spike data more intuitive, modular, and readable into a `network` instance with proper tests and validation.

jasmainak replied:
Humm ... I would keep it simple for now. You can read in a dictionary of lists. The dictionary would contain three keys `gid`, `gid_type` and `spiketimes`. It sounds like this stuff should belong to the `_Cell` class but I'm sure it's nasty to save it directly and read it back. Let's stick to python built-ins until it becomes apparent if something should go to a separate class.

jasmainak replied:
thinking more, you could probably use a list of [NamedTuple](https://docs.python.org/2/library/collections.html#collections.namedtuple). Maybe that simplifies some code, I don't know ... would keep it for another PR.

blakecaldwell replied:
Yeah, a reader would be helpful. In the future, I like the idea of moving spiketime and spikegids out of the network class. I'm in favor of either keeping things as close to HNN for now or the other extreme of making a new Spike or Spikedata class.

rythorpe replied:
The namedtuple class does look super convenient for this - good call. In the short term, I'll just implement a simple read function that validates gids prior to them overwriting the spikes of a `network` instance (to be compatible with the ranges in `network.gid_dict`).

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +518,32 @@ def plot_spikes(self, ax=None, show=True):
+            Full path to the output file (.txt)
+        trial_idx : list of int
+            Indices of selected trials. If 'all',
+            all trials are selected.
```
can you also document what the file actually looks like?

rythorpe replied:
The file is formatted in three columns where each row is <spike_time>\t<spike_gid>\t<gid_type>. There is an example in the demo.zip folder. Are you looking for a formal description in the function documentation?

jasmainak replied:
yep, a formal description in the function docstring

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +518,32 @@ def plot_spikes(self, ax=None, show=True):
+
+        spiketimes = []
+        spikegids = []
+        for i in trial_idx:
```
avoid use of single letter variable names when possible. They are hard to Ctrl + F

blakecaldwell commented:
Thanks @rythorpe!

jasmainak commented on hnn_core/dipole.py:
```diff
@@ -275,5 +275,5 @@ def write(self, fname):
         """
         X = np.r_[[self.t, self.dpl['agg'], self.dpl['L2'], self.dpl['L5']]].T
-        np.savetxt('dpl2.txt', X, fmt=['%3.3f', '%5.4f', '%5.4f', '%5.4f'],
+        np.savetxt(fname, X, fmt=['%3.3f', '%5.4f', '%5.4f', '%5.4f'],
```
can you also add a test for this? Something like:

```py
assert op.exists(fname)
```

This way we won't have a regression of the bug

jasmainak commented:
@rythorpe also don't forget to update `whats_new.rst` and `api.rst` once you have added the reader function.

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +518,32 @@ def plot_spikes(self, ax=None, show=True):
+            Indices of selected trials. If None,
+            all trials are selected.
+        """
+        if trial_idx==None:
```
```suggestion
        if trial_idx is None:
```

jasmainak replied:
https://docs.quantifiedcode.com/python-anti-patterns/readability/comparison_to_none.html

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +518,32 @@ def plot_spikes(self, ax=None, show=True):
+            spikegids += self.spikegids[idx]
+
+        gidtypes = np.empty_like(spikegids,dtype='<U36')
+        for spike_type,gid_range in self.gid_dict.items():
```
you need a PEP8 checker. For some reason Travis didn't run on your branch and PEP8 didn't run. Let me rebase your branch so you see the problem.

jasmainak replied:
That's weird your branch is up to date. @blakecaldwell any idea why Travis is not running? I saw you do some recent experiments on HNN with Travis.

blakecaldwell replied:
I'm not sure... Do you have this repository on your Travis dashboard? I haven't linked it to my account. I've found the Travis dashboard to have useful hints under the "requests" section. If it fails because of a bad .travis.yml, for example, it will indicate it there (doesn't appear to the case here, though).

jasmainak replied:
I do have it under the dashboard. Maybe @rythorpe just try pushing another commit and it might work ...

jasmainak commented:
@rythorpe do you see an error when pushing? what's the issue?

rythorpe commented:
> @rythorpe do you see an error when pushing? what's the issue?

My latest [commit](https://github.com/rythorpe/hnn-core/commit/a1d421988ac0ac773f8c4fa8801e89dd031f8d73) doesn't appear in this thread. Am I missing something?

jasmainak commented:
@rythorpe that was so strange. I'm guessing it was related to [this incident](https://www.githubstatus.com/incidents/80d0cs6kpsps). I fixed it for you now by making a new commit, pushing again, and then removing that commit.

rythorpe commented:
I logged the following error traceback when trying to update the documentation with `$ make html` from within the hnn-core/doc/ directory. Any idea what is going on?

```# Sphinx version: 2.4.4
# Python version: 3.7.7 (CPython)
# Docutils version: 0.16 release
# Jinja2 version: 2.11.1
# Last messages:

# Loaded extensions:
Traceback (most recent call last):
  File "/home/ryan/anaconda3/envs/hnn_core/lib/python3.7/site-packages/sphinx/cmd/build.py", line 275, in build_main
    args.tags, args.verbosity, args.jobs, args.keep_going)
  File "/home/ryan/anaconda3/envs/hnn_core/lib/python3.7/site-packages/sphinx/application.py", line 278, in __init__
    self._init_builder()
  File "/home/ryan/anaconda3/envs/hnn_core/lib/python3.7/site-packages/sphinx/application.py", line 334, in _init_builder
    self.events.emit('builder-inited')
  File "/home/ryan/anaconda3/envs/hnn_core/lib/python3.7/site-packages/sphinx/events.py", line 99, in emit
    results.append(callback(self.app, *args))
  File "/home/ryan/anaconda3/envs/hnn_core/lib/python3.7/site-packages/sphinx_gallery/gen_gallery.py", line 291, in generate_gallery_rst
    gallery_conf = parse_config(app)
  File "/home/ryan/anaconda3/envs/hnn_core/lib/python3.7/site-packages/sphinx_gallery/gen_gallery.py", line 91, in parse_config
    abort_on_example_error, lang, app.builder.name, app)
  File "/home/ryan/anaconda3/envs/hnn_core/lib/python3.7/site-packages/sphinx_gallery/gen_gallery.py", line 215, in _complete_gallery_conf
    "found type %s" % type(backref))
ValueError: The 'backreferences_dir' parameter must be of type str, pathlib.Path or None, found type <class 'bool'>
```

jasmainak commented:
I think we need to update that in `conf.py`. There have been some changes in `sphinx-gallery` ...

jasmainak commented:
You need to rebase @rythorpe ? Are you able to do it yourself?

Also -- I think we need to discuss this PR in person. Maybe on Thursday if you have time?

rythorpe commented:
> You need to rebase @rythorpe ? Are you able to do it yourself?
> 
> Also -- I think we need to discuss this PR in person. Maybe on Thursday if you have time?

I can rebase, though I don't understand why? Sure, Thursday works for me.

jasmainak commented:
You need to rebase because there are conflicts now. You edited the same file in both the pull requests.

jasmainak commented:
Please make a back up of your branch before you begin rebasing. It might be a mess if you are trying the first time

rythorpe commented:
@jasmainak Any idea what is going wrong with the circleci error?

jasmainak commented:
Ignore it for now, I have to configure CircleCI ...

jasmainak commented on hnn_core/__init__.py:
```diff
@@ -2,7 +2,7 @@
 load_custom_mechanisms()
 
-from .dipole import simulate_dipole
+from .dipole import simulate_dipole, import_dipole
```
```suggestion
from .dipole import simulate_dipole, read_dipole
```

jasmainak replied:
also update `api.rst`

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +522,76 @@ def plot_spikes(self, ax=None, show=True):
+            (i.e., as a trial) to the spike-related
+            attributes of Network instance. If False,
+            all spikes of the Network instance will
+            be overwritten.
```
Returns?

jasmainak commented:
@rythorpe looks like you need to rebase

jasmainak commented on hnn_core/network.py:
```diff
@@ -518,3 +522,76 @@ def plot_spikes(self, ax=None, show=True):
+                f.write('{:.3f}\t{}\t{}\n'.format(spiketimes[spk_idx],
+                        int(spikegids[spk_idx]), gidtypes[spk_idx]))
+
+    def read_spikes(self, fname, append_trial=True):
```
```suggestion
    def read_spikes(self, fname):
```

just allow `fname` to have a wildcard expression and loop over all the files that match that expression and return a list.

jasmainak commented:
An example also needs to be updated as we discussed!

rythorpe commented:
> An example also needs to be updated as we discussed!

@jasmainak yes, sorry - I wanted to run a few tests before pushing my most recent commits where I address the remaining items we discussed. I'll ping you when I push this afternoon.

jasmainak commented:
okay sounds good!

rythorpe commented:
@jasmainak I've created the `Spikes` class within network.py and integrated its functionality into `Network`, the examples (specifically, look at plot_simulate_evoked.py), and test_network.py. Let me know what you think.

jasmainak commented on hnn_core/network.py:
```diff
@@ -14,6 +14,54 @@
+    spikes : Spikes object
+        An instance of the Spikes object.
+    """
+    from glob import glob
```
You should always prefer global imports when writing a package. The reason is that the import will happen every time you call this function. So imagine if there is a function which calls this function 100 times, you will have `glob` imported 100 times and each of those imports will take some time. The only exceptions to this rule are:

1) matplotlib imports. We do these nested inside functions because of some weird bug. I don't exactly know the reason.
2) circular imports. If you file A containing function a1 calls file B containing function b, and then function b calls a function a2 in file A. Then you would need to nest the function. Usually this is an indicator of bad code structure though.
3) partial dependencies: if you want to have a partial dependency on a package so that it's imported only by users who want to use an exotic function. This should be used sparingly.

jasmainak commented on hnn_core/network.py:
```diff
@@ -14,6 +14,54 @@
 from .params import create_pext
 
 
+def read_spikes(fname, gid_dict=None):
```
This function should be added to the `api.rst` page: https://45-168215891-gh.circle-artifacts.com/0/html/api.html. Maybe add a subsection called input/output?

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
-    def plot_spikes(self, ax=None, show=True):
-        """Plot the spiking activity for each cell type.
+
+class Spikes(object):
```
can you add it to the public API here: https://45-168215891-gh.circle-artifacts.com/0/html/api.html# ? We are not yet finalized as we haven't made a release but I think it's nice to have it there so we have an overview of the documentation

jasmainak replied:
Starting from Python 3, you don't need to inherit from `object`. But I guess we have done this elsewhere, so let's remove it in another pull request

jasmainak commented on hnn_core/network.py:
```diff
@@ -525,25 +664,53 @@ def plot_spikes(self, ax=None, show=True):
+
+    def write(self, fname):
+        """Write spiking activity to a collection of spike trial
+        files.
```
This should be a single line. Quoting [PEP 257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings):

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description.

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+
+    Parameters
+    ----------
+    times : tuple (n_trials, ) of list of float | None
```
I think the code below will simplify if we use `np.ndarray` instead of tuples for `times`, `gids`, and `types`.

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+        self.times = times
+        if times is None:
+            self.times = h.Vector()
+        self.gids = gids
+        if gids is None:
+            self.gids = h.Vector()
```
```suggestion
        if gids is None:
            gids = h.Vector()
        self.gids = gids
```

if this works, I feel it's a bit easier to read. Can you also do the same for `times`

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -28,3 +28,11 @@ def test_network():
         type_key = ev_input[2: -2] + ev_input[-1]
         assert len(net.gid_dict[type_key]) == net.n_cells
+
+    # Test read/write spikes consistency
```
I think you need to add a function called `test_spikes` and check each of the methods. Here you haven't tested the `__repr__` for instance. See [how we test repr in MNE](https://github.com/mne-tools/mne-python/blob/ddc007f44c5eb840fcfd16dd3987b6535bed9f44/mne/forward/tests/test_forward.py#L55-L59)

jasmainak commented on examples/plot_simulate_evoked.py:
```diff
@@ -42,12 +42,18 @@
+# Note that we can use formatting syntax to specify the filename pattern
+# with which each trial will be written. To read spikes back in, we can use
+# wildcard expressions.
+net.spikes.plot()
```
```suggestion
```
I think it's fine to have just one plot

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+        Each gid corresponds to a type via Network::gid_dict.
+
+    Methods
+    -------
```
I have never seen this before. I looked up the [numpy docstring convention](https://numpydoc.readthedocs.io/en/latest/format.html). They say it's not really necessary but it's fine to have it. Could you format it to closely resemble it though?

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+    def __init__(self, times=None, gids=None, types=None):
+        from neuron import h
+
+        self.times = times
```
you need to do some input validation. Check that the shapes are consistent etc. Something like [this](https://github.com/mne-tools/mne-python/blob/master/mne/epochs.py#L403-L406)

jasmainak commented on hnn_core/network.py:
```diff
@@ -429,7 +477,7 @@ def _record_spikes(self):
         for gid in self._gid_list:
             if pc.gid_exists(gid):
-                pc.spike_record(gid, self.spiketimes, self.spikegids)
+                pc.spike_record(gid, self.spikes.times, self.spikes.gids)
```
how about `spikes.record(pc)`? That way you can keep everything (`times`, `gids` etc) as simple objects, e.g., numpy array and only in this method you instantiate `h.Vector()` ?

jasmainak replied:
actually never mind this comment, I'm not so sure of this

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+    times : tuple (n_trials, ) of list of float | None
+        Each element of the tuple is a trial.
+        The list contains the time stamps of spikes.
+    gids : tuple (n_trials, ) of list of float | None
```
oh wait, now I understand why you don't use numpy array. It's because each trial has different number of spikes? I think it would still be good to use list of list instead of tuple of list. Tuples are immutable. You'd use them when you don't expect to update the array

jasmainak replied:
Thinking a bit more about it, do we even need the "list of list". Why not just a flat list? Do we ever need to know which spikes belong to which trials?

rythorpe replied:
> oh wait, now I understand why you don't use numpy array. It's because each trial has different number of spikes? I think it would still be good to use list of list instead of tuple of list. Tuples are immutable. You'd use them when you don't expect to update the array

I like that spike data is stored in tuples because it makes it clear that spike data should only be modified by the `Network` class and should thus remain immutable to the user. The trade-off is that this makes our code a little less sleek. If you prefer otherwise, however, I'm happy to change it to a list of lists.

>Thinking a bit more about it, do we even need the "list of list". Why not just a flat list? Do we ever need to know which spikes belong to which trials?

HNN-gui currently plots spiking activity according to trial so that the user can visualize both aggregate and single-trial spikes. This is reflected in the current HNN I/O convention to write spikes to files on a single-trial basis. Furthermore, if we were ever to attempt running stats on different spiking states that emerge across trials, it would be really important to distinguish between trials.

jasmainak replied:
> I like that spike data is stored in tuples because it makes it clear that spike data should only be modified by the Network class and should thus remain immutable to the user.

I see, the way you would do this then is through a private attribute e.g., `_times` and then use a "property" decorator on a method called "times", so the user can only access it but not modify it (since they are not supposed to touch private attributes that start with an underscore). See [here](https://github.com/mne-tools/mne-python/blob/master/mne/epochs.py#L1452-L1455) for an example. I feel mixing tuples and lists might confuse someone reading the code in the future.

jasmainak replied:
> HNN-gui currently plots spiking activity according to trial so that the user can visualize both aggregate and single-trial spikes. This is reflected in the current HNN I/O convention to write spikes to files on a single-trial basis. Furthermore, if we were ever to attempt running stats on different spiking states that emerge across trials, it would be really important to distinguish between trials.

okay fair enough!

rythorpe replied:
Would we then need to add a `set_times()` function so that `Network` can modify the `Spikes::times` attribute?

jasmainak replied:
umm ... I don't think you need it. You are allowed to modify private attributes in the hnn-core code. But the user is not supposed to touch it. A `set_times` might be useful if it's non-trivial to change the `spikes.times` -- e.g., if it's linked to other attributes so to change it safely to add some logic.

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+
+    def __eq__(self, other):
+        if not isinstance(other, Spikes):
+            return NotImplemented
```
just raise a `ValueError` saying you can't compare `spikes` object to an object that's of a different type

jasmainak replied:
or maybe `TypeError`

rythorpe replied:
Based on my limited understanding of special methods, returning `NotImplemented` allows the interpreter to run `other.__eq__(self)` if `self.__eq__(other)` fails.

jasmainak replied:
So, if I do `Spikes(...) == 5` will it return `False` ? Is that how this will work?

This part is not tested is it?

rythorpe replied:
Correct. A return of `False` is not shown in the tests, but I have tried it on my own and it works. Should I add this as a test?

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+        times_self = [[round(time, 3) for time in trial]
+                      for trial in self.times]
+        times_other = [[round(time, 3) for time in trial]
+                       for trial in other.times]
```
Again you need some input validation. Before you check equality, you need to ensure that the shapes are consistent

jasmainak commented on hnn_core/network.py:
```diff
@@ -14,6 +14,54 @@
+        if spike_trial.shape[1] == 3:
+            spiketypes += (list(spike_trial[:, 2].astype(str)),)
+        else:
+            assert gid_dict is not None, ("Error: gid_dict must be provided "
+                                          "if spike types are unspecified in "
+                                          "'spk.txt' file")
```
the right way to raise errors is:

```suggestion
                                          "'spk.txt' file")
```

jasmainak replied:
The right way to raise errors is:

```suggestion
            if gid_dict is None:
                raise ValueError('gid_dict must be provided blah blah')
```

then you should add a test to check if these errors are actually being tested You can use `pytest.raises`

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        num_trials = len(self.times)
```
```suggestion
        n_trials = len(self.times)
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +557,99 @@ def plot_input(self, ax=None, show=True):
+        """
+
+        spiketypes = ()
+        for trl_idx in range(len(self.times)):
```
```suggestion
        for trial_idx in range(len(self.times)):
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -525,25 +664,53 @@ def plot_spikes(self, ax=None, show=True):
+
+        for trial_idx in range(len(self.times)):
+            with open(fname % (trial_idx,), 'w') as f:
+                for spk_idx in range(len(self.times[trial_idx])):
```
```suggestion
                for spike_idx in range(len(self.times[trial_idx])):
```

jasmainak commented:
@rythorpe I did a pass on comments. Most of them should be pretty easy to address. Let me know if you get stuck!

rythorpe commented:
@jasmainak I've incorporated input validation for spikes-related functions, added a bunch more tests, and made the `Spikes` attributes private along with a bunch of small cosmetic changes per your recommendations. As soon as we pass these most recent changes I'll update `api.rst` and `whats_new.rst`.

jasmainak commented on examples/plot_simulate_evoked.py:
```diff
@@ -42,12 +42,17 @@
+# with which each trial will be written. To read spikes back in, we can use
+# wildcard expressions.
+net.spikes.write('spk_%d.txt')
+spikes = read_spikes('spk_*.txt')
```
could we write this to a temporary directory to avoid messing the current directory that the user is in? You can use something here: https://docs.python.org/3/library/tempfile.html. Maybe `mkdtemp`

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -28,3 +31,51 @@ def test_network():
+
+    # TypeError should be raised when one of the args is entered as a list of
+    # non-lists (e.g., when times is a list of ints)
+    with pytest.raises(TypeError) as excinfo:
```
instead of creating `excinfo` you can use the argument `'match'` in `pytest.raises`. I think it's a cleaner way to do it.

jasmainak commented on hnn_core/network.py:
```diff
@@ -523,27 +694,54 @@ def plot_spikes(self, ax=None, show=True):
+
+        Outputs
+        -------
+        A numerically labelled txt file for each trial
```
```suggestion
        A tab separated txt file for each trial
```

These kind of files are usually called tsv files: https://en.wikipedia.org/wiki/Tab-separated_values

jasmainak commented on hnn_core/network.py:
```diff
@@ -523,27 +694,54 @@ def plot_spikes(self, ax=None, show=True):
         return ax.get_figure()
+
+    def write(self, fname):
+        """Write spiking activity to a collection of spike trial files.
```
```suggestion
        """Write spiking activity per trial to a collection of files.
```
Maybe clearer?

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+    Attributes
+    ----------
+    times : list [n_trials, ] of list of float
+        Each element of the 1st-order list is a trial.
```
```suggestion
        Each element of the outer list is a trial.
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+    ----------
+    times : list [n_trials, ] of list of float
+        Each element of the 1st-order list is a trial.
+        The 2nd-order list contains the time stamps of spikes.
```
```suggestion
        The inner list contains the time stamps of spikes.
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+
+    Parameters
+    ----------
+    times : list [n_trials, ] of list of float | None
```
```suggestion
    times : list (n_trials, ) of list (n_spikes) of float, shape | None
```

size is usually a tuple. Following numpy convention. It's immutable

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+        Each element of the 1st-order list is a trial.
+        The 2nd-order list contains the type of spike (e.g., evprox1
+        or L2_pyramidal) that occured at the corresonding time stamp.
+        Each gid corresponds to a type via Network::gid_dict.
```
```suggestion
        Each gid corresponds to a type via Network().gid_dict.
```

isn't that convention from C or C++?

rythorpe replied:
Oh yeah, my bad.

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+
+    def __init__(self, times=None, gids=None, types=None):
+        if times is None:
+            times = []
```
```suggestion
            times = list()
```

We realized some time back that it's very easy to confuse [] and (). It's better to use the full name

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+        for trial_idx in range(len(self._times)):
+            spiketypes_trial = np.empty_like(self._times[trial_idx],
+                                             dtype='<U36')
+            for gidtype, gids in gid_dict.items():
```
```suggestion
            for gid_type, gids in gid_dict.items():
```

I think it's better to put underscore so that it's clear where one word of the variable name ends and the second starts.

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+            cell or input types.
+        """
+
+        spiketypes = []
```
```suggestion
        spike_types = list()
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+            for gidtype, gids in gid_dict.items():
+                spikegids_mask = np.in1d(self._gids[trial_idx], gids)
+                spiketypes_trial[spikegids_mask] = gidtype
+            spiketypes += [list(spiketypes_trial)]
```
you can probably check by using `timeit` but I suspect this is slower than doing:
```suggestion
            spiketypes.append(spiketypes_trial)
```
also clearer. You don't need to do it the old way now that you don't have a tuple

jasmainak replied:
actually never mind this comment, I realized you need to do this because you use the `numpy` array.

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,134 @@ def plot_input(self, ax=None, show=True):
+                                             dtype='<U36')
+            for gidtype, gids in gid_dict.items():
+                spikegids_mask = np.in1d(self._gids[trial_idx], gids)
+                spiketypes_trial[spikegids_mask] = gidtype
```
what happens if some of the `gid_dict` values have overlapping ranges? Should we add a check for that?

jasmainak commented on hnn_core/network.py:
```diff
@@ -14,6 +15,52 @@
+    spikegids = []
+    spiketypes = []
+    for file in sorted(glob(fname)):
+        spike_trial = np.loadtxt(file, dtype=str)
```
what do you think of using a convention that singular is a single trial and plural is multiple trails?

So `spike` when it's just one trial and `spikes` when it's multiple trials.

rythorpe replied:
I'm afraid this will be ambiguous for others who haven't been a part of this conversation since the name of the class is `Spikes` (i.e, which includes spikes from either one or multiple trials).

jasmainak replied:
okay fair enough, we can think  of this later

jasmainak commented on hnn_core/network.py:
```diff
@@ -14,6 +15,52 @@
+    for file in sorted(glob(fname)):
+        spike_trial = np.loadtxt(file, dtype=str)
+        spiketimes += [list(spike_trial[:, 0].astype(float))]
+        spikegids += [list(spike_trial[:, 1].astype(int))]
```
```suggestion
        spike_gids += [list(spike_trial[:, 1].astype(int))]
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -14,6 +15,52 @@
+        else:
+            if gid_dict is None:
+                raise ValueError("gid_dict must be provided if spike types "
+                                 "are unspecified in 'spk.txt' file")
```
```suggestion
                                 "are unspecified in the file %s" % fname)
```

jasmainak commented:
Thanks for hanging in there @rythorpe ! I made some more comments. Mostly stylistic, so should be easy to address. We have never been closer!

jasmainak commented:
Please also update the `api.rst` so the next iteration, hopefully the PR is almost mergeable!

jasmainak commented:
@cjayb since you have been looking at the code recently, perhaps you might have some comments on this pull request?

jasmainak commented on hnn_core/network.py:
```diff
@@ -523,27 +694,54 @@ def plot_spikes(self, ax=None, show=True):
+
+        for trial_idx in range(len(self._times)):
+            with open(fname % (trial_idx,), 'w') as f:
+                for spk_idx in range(len(self._times[trial_idx])):
```
```suggestion
                for spike_idx in range(len(self._times[trial_idx])):
```

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -28,3 +31,51 @@ def test_network():
+        spikes = Spikes(times=[1, 2], gids=spikegids, types=spiketypes)
+    assert "times should be a list of lists" in str(excinfo.value)
+
+    # ValueError should be raised when one of the args is entered as an
+    # incongruent number of trials (e.g., 1 trial when all other args have 2)
```
avoid too many comments, it's bad practice (cf [Clean code](https://www.investigatii.md/uploads/resurse/Clean_Code.pdf)). Code should be written so that it can be read with comments adding only additional information.

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -28,3 +31,51 @@ def test_network():
+
+
+def test_spikes():
+    '''Test spikes object.'''
```
```suggestion
    """Test spikes object."""
```

rythorpe commented:
@jasmainak I've added `gid_dict` validation for `Spikes().update_types()` so that it checks for overlapping gid ranges. Note that I also shifted `read_spikes()` to create a `Spikes` object and then run `Spikes().update_types()` when supplied with the `gid_dict` arg (i.e., as opposed to having `read_spikes()` run its own code to populate the `Spikes()._types` attribute) to consolidate code. `api.rst` and `whats_new.rst` are updated as well.

jasmainak commented on hnn_core/network.py:
```diff
@@ -14,6 +15,52 @@
+        # provided.
+        if spike_trial.shape[1] == 3:
+            spike_types += [list(spike_trial[:, 2].astype(str))]
+        else:
+            if gid_dict is None:
```
```suggestion
        elif gid_dict is None:
```

sorry I didn't notice this before. Avoid nesting when possible. See:

```sh
$ python -c 'import this' | head -7 | tail -1
Flat is better than nested
```

rythorpe replied:
Oh good point, idk how I missed that. Any other stylistic changes I can make in this batch?

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,144 @@ def plot_input(self, ax=None, show=True):
+        if not isinstance(other, Spikes):
+            return NotImplemented
+        # Round each time element
+        times_self = [[round(time, 3) for time in trial]
+                      for trial in self._times]
+        times_other = [[round(time, 3) for time in trial]
+                       for trial in other._times]
+        return (times_self == times_other and
+                self._gids == other._gids and
+                self._types == other._types)
```
I think if you did `np.allclose(times_self, times_other, atol=1e-3) and self._types == other._types` it would do the same thing?

rythorpe replied:
I tried it in the beginning but didn't care for the fact that `np.allclose()` uses broadcasting (i.e., it won't control for the case when there has been a formatting discrepancy between `self._times` and `other._times`.

jasmainak replied:
humm ... but you won't have any formatting discrepancy now? Because you do input validation.

rythorpe replied:
There shouldn't be any formatting discrepancies due to inputs, but if a bug were to emerge in `Network` that results in a formatting discrepancy (e.g., if someone were to mess with some square brackets so that the lists append along the first dimension), it would benefit us to have an `__eq__()` function that is independent. What if we use `np.array_equal()`?

jasmainak replied:
ufff right, the Network class. Okay let's leave it like this for now. We can revisit it later.

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,144 @@ def plot_input(self, ax=None, show=True):
+            cell or input types.
+        """
+
+        # Validate gid_dict
```
```suggestion
```

jasmainak commented:
@rythorpe just two comments. Many tests added, we're going in the right direction!

LGTM once the two comments are addressed

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,144 @@ def plot_input(self, ax=None, show=True):
-        """Plot the spiking activity for each cell type.
+
+class Spikes(object):
+    '''The Spikes class.
```
```suggestion
    """The Spikes class.
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -509,8 +554,144 @@ def plot_input(self, ax=None, show=True):
+        aggregate network spiking activity according to cell type.
+    write(fname)
+        Write spiking activity to a collection of spike trial files.
+    '''
```
```suggestion
    """
```

jasmainak commented:
That's it from me, let's try to merge this PR as soon as possible so we can move on :)

cjayb commented on examples/plot_simulate_evoked.py:
```diff
@@ -42,12 +43,19 @@
+# Note that we can use formatting syntax to specify the filename pattern
+# with which each trial will be written. To read spikes back in, we can use
+# wildcard expressions.
+with tempfile.TemporaryDirectory() as tmp_dir_name:
+    print(tmp_dir_name)
+    net.spikes.write(op.join(tmp_dir_name, 'spk_%d.txt'))
+    spikes = read_spikes(op.join(tmp_dir_name, 'spk_*.txt'))
+spikes.plot()
```
I get that it's good to demo the use of the `write` method and `read_spikes` function. But I wonder if it confuses the new user (likely to look at this example first). The writing to and reading from disk are not _needed_ for plotting. Maybe both? First `net.spikes.plot()`, then the I/O-bit?

cjayb commented on hnn_core/tests/test_compare_hnn.py:
```diff
@@ -36,7 +36,7 @@ def test_hnn_core():
     # Test spike type counts
     spiketype_counts = {}
-    for spikegid in net.spikegids[0]:
+    for spikegid in net.spikes.gids[0]:
```
I'm sorry if this is obvious, but what's the difference between `spikes.gids` and `spikes._gids`?

jasmainak replied:
No difference. It's a safety mechanism so that the user doesn't modify `spikes.gids`.

cjayb commented:
L really GTM, my coding skills don't warrant deeper nit-picks.

jasmainak commented:
Thanks for taking a look!

@rythorpe the CIs are failing

rythorpe commented:
@jasmainak it turns out that there was a reason for the nested statement

rythorpe commented:
@jasmainak last thing, I just added the inputs and outputs section to `api.rst` as you had recommended earlier

rythorpe commented:
@jasmainak I think we're good to go. Do you want to move forward with the merge?

jasmainak commented:
Couple of things for future:

Inputs and Outputs should probably be formatted as a subheadheading: https://54-168215891-gh.circle-artifacts.com/0/html/api.html. Also `read_params` should be under it. We could have another subheading for cells. Anyway, future work :) I'm really done with this PR. Let's merge

jasmainak commented:
Merged, thanks @rythorpe for the contribution!
